### ISSUE TITLE
Support stablization track futures API

### ISSUE THREAD

**Issue by josephg** [bluejekyll/foundationdb-rs#111]
The API for rust futures has changed a bit since the `futures 0.1` crate, and the current version of the API is on track for stablization.

[futures-preview](https://crates.io/crates/futures-preview/0.3.0-alpha.12) exposes the new API, although internally its just re-exposing std::future. It'd be rad to support this new API in foundationdb-rs

**Comment by bluejekyll** [comment]
Thanks for publishing this. I wanted to let you know that I’m currently overcoming a boat with the flu, so won’t be able to review this at the moment. Perhaps @yjh0502 would be able to review?

**Comment by josephg** [comment]
Its just a sketch-that-compiles for now - I figured it was worth figuring out a design you're happy with before porting all the older futures code over.

But look after yourself =p - I appreciate the dedication but this can wait! Go look after yourself!

**Comment by yjh0502** [comment]
@josephg Thanks for your work!

For `impl Future`, I agree that some of those structs could be replaced with `impl Future`.

If you want to write a value after a read in a transaction, you should be able to get a current transaction from read result. You can check an example code on tests [1].

About adding `futures-preview` support on `foundationdb-rs` crate, I'm a bit worry about a stability of `futures-preview`. I tried it a few times before, but was more unstable than I expected. I thinks it's better to add `futures-preview` support after things stabilize a bit more. 

[1] https://github.com/bluejekyll/foundationdb-rs/blob/master/foundationdb/tests/get.rs#L18-L46

**Comment by josephg** [comment]
Yeah that makes sense. The problem with `futures-preview` stability is that the core team keeps making small tweaks to the API in the leadup to marking it stable. Each time the API changes, futures-preview needs retooling (which they're pretty quick about), but afterwards it only works on rust nightly versions new enough to have the API updates. This is about to happen again when they rename `LocalWaker` to `Waker`. (The std churn is because unlike futures 0.1, the future trait exported by futures-preview is just `std::future::Future`).

Once futures land in stable, futures-preview should stop breaking. (I imagine at that point it'll just become `futures-0.3`. Maybe we could make a branch which uses the new futures API, and hold off merging it into master until futures reaches stable (which will probably be pretty soon - it hit final comment period in December).

The other option is to do without any of the futures combinators, but they're very useful. I'd much rather rely on [`AtomicWaker`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.11/futures/task/struct.AtomicWaker.html) than write my own threading / signalling code to signal completion back to the appropriate thread.

As for that example, yikes! You have to `#![feature(async_await, await_macro]` but with new futures we should be able to use the library like this:

```rust
async {
    let cluster = await!(Cluster::new(foundationdb::default_config_path()))?;
    let db = await!(cluster.create_database())?;
    await!(db.do_tn(|tn| {
        tn.set(b"hello", b"world");
    }))?;

    let value = await!(db.do_tn(async |tn| {
        await!(tn.get(b"hello"))
    }))?;
    // ...
}
```

(I've been totally spoiled writing async/await code in other languages).

**Comment by josephg** [comment]
I've gotten a hello world example working. I ended up needing to make a wrapper type for txn future results because the memory is owned by the fdb future object itself. The API currently comes out looking like this:

```rust
block_on(async {
    let cluster = await!(Cluster::new(foundationdb::default_config_path())).unwrap();
    let db = await!(cluster.create_database()).unwrap();

    await!(db.transact(async move |tr| {
        tr.set(b"hello", b"world");
        Ok(())
    })).unwrap();

    let result = await!(db.transact(|tr| tr.get(b"hello", false))).unwrap();
    assert_eq!(*result.unwrap(), b"world");
}
```

The `*result` is using the `Deref` trait to grab the juicy value itself inside; which lives as long as `result`.

This also supports running all the retry logic; which is what you basically always want as a consumer of the API.

I'm tempted to suggest we should make the transaction function `Fn` instead of `FnMut` for safety, since your function might be called multiple times if there's contention.

I've also disabled range queries, watches and almost all the old tests. But with a bit of cleanup we should be able to enable this stuff again. There should be 0 allocations / unnecessary copies in rust on the normal path, although we are unfortunately incrementing and decrementing the Arc refcount of `Database` and `Cluster` a whole lot more than we need to. But this should be pretty cheap relative to the normal network overheads of the client.
